src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc

```diff
@@ -0,0 +1,79 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+#include "src/ir/auth_logic/auth_logic_ast_traversing_visitor.h"
+
+#include "absl/container/flat_hash_set.h"
+#include "src/common/testing/gtest.h"
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/datalog/program.h"
+
+namespace raksha::ir::auth_logic {
+namespace {
+
+// A visitor that makes a set of all the names of principals in the program
+class PrincipalNameCollectorVisitor
+    : public AuthLogicAstTraversingVisitor<PrincipalNameCollectorVisitor,
+                                           absl::flat_hash_set<std::string>> {
+ public:
+  PrincipalNameCollectorVisitor() {}
+
+  absl::flat_hash_set<std::string> GetDefaultValue() override { return {}; }
+
+  absl::flat_hash_set<std::string> CombineResult (
+      absl::flat_hash_set<std::string> acc,
```

Comment:
Is it correct that these should be passed by value?

Reply:
I left this open in case I am missing something though, which might be the case.

Reply:
I'm assuming this was OK and resolving.